Index: VEX/priv/guest-x86/toIR.c
===================================================================
--- VEX/priv/guest-x86/toIR.c	(revision 1633)
+++ VEX/priv/guest-x86/toIR.c	(working copy)
@@ -12644,6 +12644,15 @@ DisResult disInstr_X86_WRK ( 
          DIP("emms\n");
          break;
 
+      case 0x1F: /* NOP [HINT] */
+      {
+         int len;
+         addr = disAMode ( &len, sorb, delta, dis_buf );
+         delta +=len;
+         DIP("NOP [HINT]");
+         break;
+      }
+
       /* =-=-=-=-=-=-=-=-=- unimp2 =-=-=-=-=-=-=-=-=-=-= */
 
       default:
